src/go/intelligence/pipeline/add_annotations.go

```diff
@@ -0,0 +1,123 @@
+package pipeline
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"go/intelligence/models"
+	"go/intelligence/utils"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+)
+
+// Module     : intelligence
+// Author     : Rahul Indra <indrarahul2013 AT gmail dot com>
+// Created    : Wed, 1 July 2020 11:04:01 GMT
+// Description: CMS MONIT infrastructure Intelligence Module
+
+//AddAnnotation function for adding annotations to dashboards
+func AddAnnotation(data <-chan models.AmJSON) <-chan models.AmJSON {
+
+	dataAfterAnnotation := make(chan models.AmJSON)
+
+	go func() {
+		dashboards := utils.FindDashboard()
+		for each := range data {
```

Comment:
Please provide clear explanation of such complex structure, so far it is O(n^5) algorithm which by any definition is VERY slow. I think you should split it apart and provide O(n) algorithm. For instance, right series of for loops (rather netested one), where you need to find the following:
- find list of dashboards where annotations should be applied.
- find list of services in config
- find list of services in Annotation map
- take the intersection of aforementioned configs
- loop over final list of services and apply it to your list of dashboads.

Each loop is O(n), therefore the above logic will be N^4 times faster of your approach.

Reply:
@vkuznet  This is true. I have optimized the code. Please review